### PR TITLE
LibPartition+PartitionEditor: Add the beginnings of a partition editor

### PR DESCRIPTION
**Pull request by SamuelBowman**
This PR takes the first steps towards building a full featured partition editor for Serenity! :^)

**First, we need a way to read partition tables from userland.**
There was already a bunch of partition table parsing code in the kernel, so I extracted it into a new library called LibPartition. The code was initially dependent on Kernel::StorageDevice, but with some strategic #ifdef-ing, was made to work with a Core::File as well. The end result is a kernel/userland agnostic LibPartition capable of reading MBR, EBR, and GPT partition tables. (There's probably a better way to accomplish this with fewer #ifdefs, but this works as a first pass.)

**Then, all we need is the PartitionEditor application!**
Since LibPartition doesn't know how to write partition tables yet, PartitionEditor is currently read-only. On startup, it scans /dev for block devices and uses LibPartition to parse any partition tables on the device and display them in a table. Since block devices are chmod 600, PartitionEditor must unfortunately be run as root. A message box is displayed if you try to run PartitionEditor as a non-root user. (Ideally, we would prompt the user to elevate to superuser kind of like how FileSystemAccessServer does or even just use a second FileSystemAccessServer running as root instead of the normal one running as anon, but that's a yak for another day.)

**Screenshot:**
Screenshot_2022-06-29_22-50-09

**Testing:**
By default, Serenity uses the entire disk for its ext2 filesystem without a partition table. This isn't useful for testing as the partition list for the default disk will be empty. To test properly, I created a few [disk images](https://github.com/SerenityOS/serenity/files/9016043/disk_images.zip) and instructed qemu to attach them using the following command: `export SERENITY_EXTRA_QEMU_ARGS="-drive file=/path/to/mbr.img,format=raw,index=1,media=disk -drive file=/path/to/ebr.img,format=raw,index=2,media=disk -drive file=/path/to/gpt.img,format=raw,index=3,media=disk"`

**Next steps:**
- Draw a logo for PartitionEditor. (I just used SpaceAnalyzer's logo.)
- Add write support to LibPartition and PartitionEditor.
- General incremental improvements.

I've been working on all this on and off since (checks notes) January 16th 😬 :skeleyak: and finally got it to the point where I think it's ready for review. Once this groundwork is laid, it should be much easier to make incremental improvements through.

### CODE REVIEW

**Comment by supercomputer7** [comment]
Great work :)
I have not looked into the patch series, but the result looks very polished for this stage and time :)

This might be integrated very well with #13779, so ideally we should not look into `/dev` to get any info about partition layouts.

File: Userland/Applications/PartitionEditor/PartitionModel.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            return partition.end_block();
+        default:
+            VERIFY_NOT_REACHED();
+        }
+    }
+
+    return {};
+}
+
+ErrorOr<void> PartitionModel::set_device_path(String path)
```
Comment (linusg) [inline Userland/Applications/PartitionEditor/PartitionModel.cpp:63]: I don't think we need to take this by value if we don't move it anyway

File: Userland/Applications/PartitionEditor/main.cpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+#include <LibGUI/ItemListModel.h>
+#include <LibGUI/Menu.h>
+#include <LibGUI/MessageBox.h>
+#include <LibGUI/TableView.h>
+
+using namespace PartitionEditor;
+
+Vector<String> get_device_paths();
+
+Vector<String> get_device_paths()
```
Comment (linusg) [inline Userland/Applications/PartitionEditor/main.cpp:23]: This can just be `static`

File: Userland/Applications/PartitionEditor/main.cpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+#include <LibCore/DirIterator.h>
+#include <LibCore/System.h>
+#include <LibGUI/Application.h>
+#include <LibGUI/ComboBox.h>
+#include <LibGUI/ItemListModel.h>
+#include <LibGUI/Menu.h>
+#include <LibGUI/MessageBox.h>
+#include <LibGUI/TableView.h>
+
+using namespace PartitionEditor;
```
Comment (linusg) [inline Userland/Applications/PartitionEditor/main.cpp:19]: Please don't

File: Userland/Applications/PartitionEditor/main.cpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+    // FIXME: PartitionEditor needs its own icon.
+    auto app_icon = TRY(GUI::Icon::try_create_default_icon("app-space-analyzer"));
+
+    auto window = TRY(GUI::Window::try_create());
+    window->set_title("Partition Editor");
+    window->resize(640, 400);
+    window->set_icon(app_icon.bitmap_for_size(16));
+
+    // FIXME: Abort and show a dialog if not running as root.
```
Comment (linusg) [inline Userland/Applications/PartitionEditor/main.cpp]: That doesn't seem like great UX. Instead we should have functionality to prompt the user for a password to authenticate as root once needed (this may already exist; not sure).

...this is assuming that read-only use doesn't need root :)

File: Userland/Applications/PartitionEditor/main.cpp [hunk 5]
```diff
@@ -50,7 +51,10 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
     window->resize(640, 400);
     window->set_icon(app_icon.bitmap_for_size(16));
 
-    // FIXME: Abort and show a dialog if not running as root.
+    if (getuid() != 0) {
+        GUI::MessageBox::show_error(window, "PartitionEditor must be run as root");
+        return Error::from_string_literal("PartitionEditor must be run as root");
```
Comment (linusg) [inline Userland/Applications/PartitionEditor/main.cpp:56]: ...but why?

**Comment by SamuelBowman** [comment]
> Great work :) I have not looked into the patch series, but the result looks very polished for this stage and time :)

Thanks! I sure did take quite a while to get to this point though lol.

> This might be integrated very well with #13779, so ideally we should not look into `/dev` to get any info about partition layouts.

I took a quick look at #13779 and I'm not exactly sure what you mean here.

If you just mean we shouldn't look in /dev to discover block devices, it should be pretty trivial to make PartitionEditor discover block devices from the symlinks in /sys/dev/block instead.

If you instead mean we shouldn't use LibPartition to read partition tables from the block devices in /dev, I don't think that's a good idea. We'll ultimately need to write to the block devices using LibPartition anyways, and adding /sys/devices/storage as a second data source seems ripe for some nasty bugs where the two could differ and result in corruption on write. Using LibPartition for both read and write is simpler and avoids any potential pitfalls. To clarify, I'm not at all opposed to exposing partition info in /sys, I'm just opposed to using it as the source of truth for PartitionEditor.

If you meant something else entirely, you'll have to explain as my small brain totally missed it :^)

File: Userland/Libraries/LibPartition/CMakeLists.txt [hunk 6]
```diff
@@ -7,3 +7,4 @@ set(SOURCES
 )
 
 serenity_lib(LibPartition Partition)
+target_link_libraries(LibPartition LibC)
```
[the remark this answers is not in the document]
Comment (SamuelBowman) [reply Userland/Libraries/LibPartition/CMakeLists.txt:10]: Without LibC, it fails to link with a bunch of undefined references to __stack_chk_fail_local on the closing curly brace of method definitions. I'm assuming the references are being inserted by the compiler as part of stack protection, but I'm not super knowledgeable about that. Potentially related? #14376

File: Userland/Libraries/LibPartition/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,5 @@
+set(SOURCES
+    DiskPartitionMetadata.cpp
+)
+
+serenity_lib(LibPartition Partition)
```
[the remark this answers is not in the document]
Comment (SamuelBowman) [reply Userland/Libraries/LibPartition/CMakeLists.txt]: Fixed.

File: Userland/Libraries/LibPartition/EBRPartitionTable.h [hunk 8]
```diff
@@ -15,12 +15,25 @@ class EBRPartitionTable : public MBRPartitionTable {
...
-    virtual bool is_valid() const override { return m_valid; };
+#else
+    static ErrorOr<NonnullOwnPtr<EBRPartitionTable>> try_to_initialize(NonnullRefPtr<Core::File>);
+    explicit EBRPartitionTable(NonnullRefPtr<Core::File>);
+#endif
+
+    virtual bool is_valid() const override
+    {
+        return m_valid;
+    };
```
[the remark this answers is not in the document]
Comment (SamuelBowman) [reply Userland/Libraries/LibPartition/EBRPartitionTable.h]: Epic clang-format pre-commit fail. Fixed.

File: Userland/Applications/PartitionEditor/PartitionModel.cpp (hunk 1 above)
Comment (SamuelBowman) [reply Userland/Applications/PartitionEditor/PartitionModel.cpp]: Good catch. Fixed.

File: Userland/Applications/PartitionEditor/main.cpp (hunk 2 above)
Comment (SamuelBowman) [reply Userland/Applications/PartitionEditor/main.cpp]: TIL C++ has C style static functions outside of classes. Fixed.

File: Userland/Applications/PartitionEditor/main.cpp (hunk 3 above)
Comment (SamuelBowman) [reply Userland/Applications/PartitionEditor/main.cpp]: Fixed. For reference, I was using TextEditor as a template. Is there a reason some apps (FileManager, Help, ImageViewer, Music, TextEditor) are using their namespaces while the rest don't?

File: Userland/Applications/PartitionEditor/main.cpp (hunk 4 above)
Comment (SamuelBowman) [reply Userland/Applications/PartitionEditor/main.cpp]: I 100% agree. This is not great UX and we should prompt the user to authenticate as root. Unfortunatly, AFAIK no such functionality exists for GUI apps yet (CLI has pls). Without this message box, attempting to run as a non-root user silently fails with a permission error, which is really bad UX.

File: Userland/Applications/PartitionEditor/main.cpp (hunk 5 above)
Comment (SamuelBowman) [reply Userland/Applications/PartitionEditor/main.cpp:52]: Block devices are owned by root and chmod 600. This is necessary to prevent unprivileged users from being able to access the block device directly and bypass filesystem permissions. To read partition tables, we need access to the raw block device though.

File: Userland/Libraries/LibPartition/CMakeLists.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (linusg) [reply Userland/Libraries/LibPartition/CMakeLists.txt:10]: Seems like the completely wrong solution though, just link against LibCore like almost every other library, and not LibC directly.

File: Userland/Applications/PartitionEditor/main.cpp (hunk 5 above)
Comment (linusg) [reply Userland/Applications/PartitionEditor/main.cpp:52]: Fine, but let's put a concise summary of that into the message and not just "root pls not telling you why tho"

File: Userland/Applications/PartitionEditor/main.cpp (hunk 3 above)
Comment (linusg) [reply Userland/Applications/PartitionEditor/main.cpp]: Not sure, probably just historical - someone added it, it got missed during code review, and then served as a template for a bunch of other apps.

**Comment by supercomputer7** [comment]
> > Great work :) I have not looked into the patch series, but the result looks very polished for this stage and time :)
> 
> Thanks! I sure did take quite a while to get to this point though lol.
> 
> > This might be integrated very well with #13779, so ideally we should not look into `/dev` to get any info about partition layouts.
> 
> I took a quick look at #13779 and I'm not exactly sure what you mean here.
> 
> If you just mean we shouldn't look in /dev to discover block devices, it should be pretty trivial to make PartitionEditor discover block devices from the symlinks in /sys/dev/block instead.
> 
> If you instead mean we shouldn't use LibPartition to read partition tables from the block devices in /dev, I don't think that's a good idea. We'll ultimately need to write to the block devices using LibPartition anyways, and adding /sys/devices/storage as a second data source seems ripe for some nasty bugs where the two could differ and result in corruption on write. Using LibPartition for both read and write is simpler and avoids any potential pitfalls. To clarify, I'm not at all opposed to exposing partition info in /sys, I'm just opposed to using it as the source of truth for PartitionEditor.
> 
> If you meant something else entirely, you'll have to explain as my small brain totally missed it :^)

What I mean is that we could easily expose partition metadata in the `/sys` filesystem, without requiring any sort of root permissions for this. That PR is just one step towards that :)

File: Userland/Libraries/LibPartition/CMakeLists.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (SamuelBowman) [reply Userland/Libraries/LibPartition/CMakeLists.txt]: Done

File: Userland/Applications/PartitionEditor/main.cpp (hunk 5 above)
Comment (SamuelBowman) [reply Userland/Applications/PartitionEditor/main.cpp]: Changed to "PartitionEditor must be run as root in order to open raw block devices and read partition tables."

**Comment by SamuelBowman** [comment]
Thanks for the elaboration! I definitely missed some things the first time :^)

> What I mean is that we could easily expose partition metadata in the `/sys` filesystem, without requiring any sort of root permissions for this. That PR is just one step towards that :)

Oohh! That could be nice. I suppose I've gotten used to root being needed for anything partition table or file system related on Linux, but thinking about it, root should only be really needed to make changes. Partition layout metadata isn't sensitive information, so it's a little unfortunate to require root just to read the partition tables.

> As for write support of partition tables, Linux seems to support a special ioctl to reload the partition table after a userspace program modified the partition table of the disk (https://man7.org/linux/man-pages/man4/sd.4.html, look for `BLKRRPART`), so we should support it too, and by doing so, we can rely on the information from `/sys` because the kernel will just reload the partition table and will change the information in `/sys` accordingly.

We'll definitely want to add a reload ioctl once we start writing partition tables. `BLKRRPART` is what Linux uses, but this doesn't appear to be standard across other Unixlikes (see https://github.com/samangh/gptfdisk/blob/master/diskio-unix.cc#L255). I'd suggest something like `STORAGE_DEVICE_RESCAN_PARTITION_TABLES` which better matches our other 2 StorageDevice ioctls `STORAGE_DEVICE_GET_SIZE` and `STORAGE_DEVICE_GET_BLOCK_SIZE`.

I've also reflected a bit more on my previous concerns regarding using data from `/sys` as the source of truth for PartitionEditor, and I think those concerns could be effectively mitigated by simply `VERIFY`-ing the data from `/sys` matches the data from LibPartition before committing any changes.

---

TL;DR: I like your proposal and would happily update PartitionEditor to take advantage of it :^)

**Comment by linusg** [comment]
This has a conflict
